code/game/objects/items/devices/aicard.dm

```diff
@@ -38,102 +37,54 @@
-			var/law = A.laws.supplied[index]
-			if (length(law) > 0)
-				laws += "[number]: [law]<BR>"
-				number++
-
-		dat += "Laws:<br>[laws]<br>"
-
-		if (A.stat == 2)
-			dat += "<b>AI nonfunctional</b>"
-		else
-			if (!src.flush)
-				dat += {"<A href='byond://?src=\ref[src];choice=Wipe'>Wipe AI</A>"}
-			else
-				dat += "<b>Wipe in progress</b>"
-			dat += "<br>"
-			dat += {"<a href='byond://?src=\ref[src];choice=Wireless'>[A.control_disabled ? "Enable" : "Disable"] Wireless Activity</a>"}
-			dat += "<br>"
-			dat += {"<a href='byond://?src=\ref[src];choice=Radio'>[A.radio_enabled ? "Disable" : "Enable"] Subspace Radio</a>"}
-			dat += "<br>"
-			dat += {"<a href='byond://?src=\ref[src];choice=Close'>Close</a>"}
-	user << browse(dat, "window=aicard")
-	onclose(user, "aicard")
-	return
 
-/obj/item/device/aicard/Topic(href, href_list)
-	var/mob/U = usr
-	if (!in_range(src, U)||U.machine!=src)//If they are not in range of 1 or less or their machine is not the card (ie, clicked on something else).
-		U << browse(null, "window=aicard")
-		U.unset_machine()
-		return
-
-	add_fingerprint(U)
-	U.set_machine(src)
-
-	switch(href_list["choice"])//Now we switch based on choice.
-		if ("Close")
-			U << browse(null, "window=aicard")
-			U.unset_machine()
-			return
-
-		if ("Wipe")
+/obj/item/device/aicard/get_ui_data()
+	var/list/data = list()
+	if(AI)
+		data["name"] = AI.name
+		data["laws"] = AI.laws.get_law_list(include_zeroth=1)
+		data["health"] = (AI.health+100)/2
+		data["wireless"] = !AI.control_disabled //todo disabled->enabled
+		data["radio"] = AI.radio_enabled
+		data["isDead"] = AI.stat == DEAD
+		data["isBraindead"] = AI.client ? TRUE : FALSE
+	data["wiping"] = flush
+	return data
+
+/obj/item/device/aicard/ui_act(action,params)
+	if(..())
+		return 1
+
+	switch(action)
+		if("wipe")
 			var/confirm = alert("Are you sure you want to wipe this card's memory? This cannot be undone once started.", "Confirm Wipe", "Yes", "No")
 			if(confirm == "Yes")
-				if(isnull(src)||!in_range(src, U)||U.machine!=src)
-					U << browse(null, "window=aicard")
-					U.unset_machine()
-					return
-				else
-					flush = 1
-					for(var/mob/living/silicon/ai/A in src)
-						A.suiciding = 1
-						A << "Your core files are being wiped!"
-						while (A.stat != 2)
-							A.adjustOxyLoss(2)
-							A.updatehealth()
-							sleep(10)
-						flush = 0
-
-		if ("Wireless")
-			for(var/mob/living/silicon/ai/A in src)
-				A.control_disabled = !A.control_disabled
-				A << "The intellicard's wireless port has been [A.control_disabled ? "disabled" : "enabled"]!"
-				if (A.control_disabled)
-					overlays -= image('icons/obj/aicards.dmi', "aicard-on")
-				else
-					overlays += image('icons/obj/aicards.dmi', "aicard-on")
-
-		if ("Radio")
-			for(var/mob/living/silicon/ai/A in src)
-				A.radio_enabled = !A.radio_enabled
-				A << "Your Subspace Transceiver has been [A.radio_enabled ? "enabled" : "disabled"]!"
-	attack_self(U)
+				.("wipe_confirm")

```

Comment:
except you could just put that in wipe, instead of adding an entire new topic section? that's what I'm complaining about, having topic call itself needlessly.